rcl/src/rcl/client.c

```diff
@@ -153,6 +171,10 @@ rcl_client_init(
     remapped_service_name);
   goto cleanup;
 fail:
+  if (client->impl->service_event_publisher) {
+    allocator->deallocate(client->impl->service_event_publisher, allocator->state);
+    client->impl->service_event_publisher = NULL;
+  }
```

Comment:
This check should happen inside the `if (client->impl)` block below, since `client->impl` may be null.